src/audio/ipcgtw.c

```diff
@@ -0,0 +1,397 @@
+
+	comp_set_drvdata(dev, ipcgtw_data);
+
+	ipcgtw_data->node_id = gtw_cfg->node_id;
+	ipcgtw_data->dev = dev;
+
+	if (!gtw_cfg->config_length) {
+		comp_cl_err(&comp_ipcgtw, "ipcgtw_new(): empty ipc4_gateway_config_data");
+		goto error;
+	}
+
+	blob = (const struct ipc4_ipc_gateway_config_blob *)
+		((const struct ipc4_gateway_config_data *)gtw_cfg->config_data)->config_blob;
+
+	/* Endpoint buffer is created in copier with size specified in copier config. That buffer
+	 * will be resized to size specified in IPC gateway blob later in ipcgtw_params().
+	 */
+	comp_cl_dbg(&comp_ipcgtw, "ipcgtw_new(): buffer_size: %u", blob->buffer_size);
+	ipcgtw_data->buf_size = blob->buffer_size;
+
+	if (!ipcgtw_list_head.next)
+		list_init(&ipcgtw_list_head);
+
+	list_init(&ipcgtw_data->item);
+	list_item_append(&ipcgtw_data->item, &ipcgtw_list_head);
+
+	dev->state = COMP_STATE_READY;
+	return dev;
+
+error:
+	rfree(ipcgtw_data);
+	rfree(dev);
+	return NULL;
+}
+
+static void ipcgtw_free(struct comp_dev *dev)
+{
+	struct ipcgtw_data *ipcgtw_data = comp_get_drvdata(dev);
+
+	list_item_del(&ipcgtw_data->item);
+
+	rfree(ipcgtw_data);
+	rfree(dev);
+}
+
+static struct comp_dev *find_ipcgtw_by_node_id(union ipc4_connector_node_id node_id)
+{
+	struct list_item *item;
+
+	list_for_item(item, &ipcgtw_list_head) {
+		struct ipcgtw_data *data = list_item(item, struct ipcgtw_data, item);
+
+		if (data->node_id.dw == node_id.dw)
+			return data->dev;
+	}
+
+	return NULL;
+}
+
+static inline void audio_stream_copy_bytes_from_linear(const void *linear_source,
+						       struct audio_stream __sparse_cache *sink,
+						       unsigned int bytes)
+{
+	const uint8_t *src = (const uint8_t *)linear_source;
+	uint8_t *snk = audio_stream_wrap(sink, (uint8_t *)sink->w_ptr);
+	size_t bytes_snk, bytes_copied;
+
+	while (bytes) {
+		bytes_snk = audio_stream_bytes_without_wrap(sink, snk);
+		bytes_copied = MIN(bytes, bytes_snk);
+		memcpy_s(snk, bytes_copied, src, bytes_copied);
+		bytes -= bytes_copied;
+		src += bytes_copied;
+		snk = audio_stream_wrap(sink, snk + bytes_copied);
+	}
+}
+
+static inline
+void audio_stream_copy_bytes_to_linear(const struct audio_stream __sparse_cache *source,
+				       void *linear_sink, unsigned int bytes)
+{
+	uint8_t *src = audio_stream_wrap(source, (uint8_t *)source->r_ptr);
+	uint8_t *snk = (uint8_t *)linear_sink;
+	size_t bytes_src, bytes_copied;
+
+	while (bytes) {
+		bytes_src = audio_stream_bytes_without_wrap(source, src);
+		bytes_copied = MIN(bytes, bytes_src);
+		memcpy_s(snk, bytes_copied, src, bytes_copied);
+		bytes -= bytes_copied;
+		src = audio_stream_wrap(source, src + bytes_copied);
+		snk += bytes_copied;
+	}
+}
+
+static inline struct comp_buffer *get_buffer(struct comp_dev *dev)
+{
+	if (dev->direction == SOF_IPC_STREAM_PLAYBACK) {
+		if (list_is_empty(&dev->bsink_list))
+			return NULL;
+		return list_first_item(&dev->bsink_list, struct comp_buffer, source_list);
+	}
+
+	assert(dev->direction == SOF_IPC_STREAM_CAPTURE);
```

Comment:
Not related directly to this PR, but @marcinszkudlinski and @aborisovich , here's one good example for the discussion whether to enable asserts by default in all builds.

- If asserts are not enabled in production builds, this is completely ok and I wouldn't think about it more in a review.

- If asserts are always enabled, then this starts to be more complicated. There's no logic directly on the direction in this function, and this is inline helper so can be called on potentially hot codepaths, so not sure this is needed.

This specific case seems ok in the end (as it's not called from any audio loops directly), but a good example that if we do enable assert across the board, then that will have implications to their usage.

Reply:
Yes and no
Asserts should be used in 2 cases
1) when there's absolutely no way to recover - controlled system panic is better than crash
2) in situations that "not gonna happen" - in this particular case the field may get 2 values, we really need to panic if it's set to 3 cause it is a defect (memory corruption?) the sooner we know about it the better. 

And IMHO both cases should be enabled in production

Completely separate matter is performance impact of asserts. 

IDEA - maybe 2 levels of assets? Debug asserts and "hard asserts" ? First ones disabled in production?

Reply:
@marcinszkudlinski Ack on 1+2 but the problem is specifically on the performance aspects. For (2) type of checks, there's no upper limit of how paranoid you can get. But let's continue the discussion in other PRs related to asserts, I just wanted to raise this specific point as a good example to think about.

Reply:
The name `assert` is already used and defined by the C standard, the Zephyr documentation and probably others. They both clearly expect assert to be **disabled in "production"**:
https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/assert.h.html
https://docs.zephyrproject.org/1.14.0/reference/kconfig/CONFIG_ASSERT.html

HOWEVER they don't define what "production" is. Is Continuous Integration on the main branch production? Unlike a release branch, I would say Continuous Integration on the main branch is NOT production.

> Ack on 1+2 but the problem is specifically on the performance aspects.

Exactly: the only reason to disable asserts is to save memory or CPU or both = performance. I think "production" implies "performance". So a "performance" build configured for benchmarks and detecting regressions should also disable asserts... even on the main branch!

So the answer to "enabled by default?" question is: BOTH. It depends what gets tested, what is the purpose of the build.

> IDEA - maybe 2 levels of assets? Debug asserts and "hard asserts" ?


Good idea. `ALWAYS_ASSERT()`?